internal/vm/vsock.go

```diff
@@ -144,23 +228,93 @@ func VSockAcceptConnector(port uint32) IOConnector {
+	// The message a guest-side connection will write after accepting a connection from
+	// a host dial. This is not part of the official Firecracker vsock spec, but is
+	// recommended in order to allow the host to verify connections were established
+	// successfully: https://github.com/firecracker-microvm/firecracker/issues/1272#issuecomment-533004066
+	return fmt.Sprintf("IMALIVE %d\n", port)
+}
+
+// tryConnRead will try to do a read from the provided conn, returning an error if
+// the bytes read does not match what was provided or if the read does not complete
+// within the provided timeout. It will reset socket deadlines to none after returning.
+func tryConnRead(conn net.Conn, expectedRead string, timeout time.Duration) error {
+	conn.SetDeadline(time.Now().Add(timeout))
+	defer conn.SetDeadline(time.Time{})
+
+	actualRead := make([]byte, len(expectedRead))
+	_, err := conn.Read(actualRead)
+	if err != nil {
+		return err
+	}
+
+	if expectedRead != string(actualRead) {
+		return errors.Errorf("expected to read %q, but instead read %q",
+			expectedRead, string(actualRead))
+	}
+
+	return nil
+}
+
+// tryConnWrite will try to do a write to the provided conn, returning an error if
+// the write fails or does not complete within the provided timeout. It will reset
+// socket deadlines to none after returning.
+func tryConnWrite(conn net.Conn, expectedWrite string, timeout time.Duration) error {
+	conn.SetDeadline(time.Now().Add(timeout))
+	defer conn.SetDeadline(time.Time{})
+
+	bytesWritten, err := conn.Write([]byte(expectedWrite))
+	if err != nil {
+		return err
+	}
+	if bytesWritten != len(expectedWrite) {
+		return errors.Errorf("incomplete write, expected %d bytes but wrote %d",
+			len(expectedWrite), bytesWritten)
+	}
+
+	return nil
+}
+
+type vsockConnectMsgError struct {
+	cause error
 }
 
-// Unfortunately, as "documented" on various online forums, there's no ideal way to
-// test for actual Linux error codes returned by the net library or wrappers
-// around that library. The common approach is to fall back on string matching,
-// which is done for the functions below
+func (e vsockConnectMsgError) Error() string {
+	return errors.Wrap(e.cause, "vsock connect message failure").Error()
+}
+
+func (e vsockConnectMsgError) Temporary() bool {
+	return false
+}
+
+type vsockAckError struct {
+	cause error
+}
+
+func (e vsockAckError) Error() string {
+	return errors.Wrap(e.cause, "vsock ack message failure").Error()
+}
 
-func isENXIO(err error) bool {
-	return strings.HasSuffix(err.Error(), "no such device")
+func (e vsockAckError) Temporary() bool {
+	return true
 }
 
-func isECONNRESET(err error) bool {
-	return strings.HasSuffix(err.Error(), "connection reset by peer")
+// isTemporaryNetErr returns whether the provided error is a retriable
+// error, according to the interface defined here:
+// https://golang.org/pkg/net/#Error
+func isTemporaryNetErr(err error) bool {
+	terr, ok := err.(interface {
```

Comment:
Any reason not to do this:

```go
type tempErr interface {
	Temporary() bool
}

func isTemporaryNetErr(err error) bool {
	terr, ok := err.(tempErr)
	return err != nil && ok && terr.Temporary()
}
```

Reply:
Didn't realize it was pre-existing, feel free to ignore.  And yeah, I don't think it's massively different except for avoiding an inline interface and having a named interface instead.